ibis/backends/dask/udf.py

```diff
@@ -0,0 +1,161 @@
+        return op.func(*args)
+
+    return scope
+
+
+@pre_execute.register(ops.AnalyticVectorizedUDF)
+@pre_execute.register(ops.AnalyticVectorizedUDF, ibis.client.Client)
+@pre_execute.register(ops.ReductionVectorizedUDF)
+@pre_execute.register(ops.ReductionVectorizedUDF, ibis.client.Client)
+def pre_execute_analytic_and_reduction_udf(op, *clients, scope=None, **kwargs):
+    input_type = op.input_type
+    nargs = len(input_type)
+
+    # An execution rule to handle analytic and reduction UDFs over
+    # 1) an ungrouped window,
+    # 2) an ungrouped Aggregate node, or
+    # 3) an ungrouped custom aggregation context
+    @execute_node.register(type(op), *(itertools.repeat(dd.Series, nargs)))
+    def execute_udaf_node_no_groupby(op, *args, aggcontext, **kwargs):
+
+        meta = pandas.Series([], dtype=op._output_type.to_dask())
+
+        @dask.delayed
+        def lazy_agg(*series):
+            return aggcontext.agg(series[0], op.func, *series[1:])
+
+        lazy_result = lazy_agg(*args)
+        # Depending on the type of operation, lazy_result is a Delayed that
+        # could become a dd.Series or a dd.core.Scalar
+        if isinstance(op, ops.AnalyticVectorizedUDF):
+            result = dd.from_delayed(lazy_result, meta=meta)
+        else:
+            # lazy_result is a dd.core.Scalar from an ungrouped reduction
+            # computation is forced so that we can work with the scalar
+            # TODO computation
+            out_dtype = op._output_type.to_dask()
+            if out_dtype == object:
+                # we're outputing a dt.Struct that will need to be destructured
+                # we compute so we can work with scalars downstream
+                result = lazy_result.compute()
+            else:
+                result = dd.from_delayed(
+                    lazy_result, meta=out_dtype, verify_meta=False
+                )
+
+        return result
+
+    @execute_node.register(
+        type(op), *(itertools.repeat(ddgb.SeriesGroupBy, nargs))
+    )
+    def execute_udaf_node_groupby(op, *args, aggcontext, **kwargs):
+        # To apply a udf func to a list of grouped series we:
+        # 1. Grab the dataframe they're grouped off of
+        # 2. Grab the column name for each series
+        # 3. .apply a wrapper that performs the selection using the col name
+        #    and applies the udf on to those
+        # This way we rely on dask dealing with groups and pass the udf down
+        # to the frame level.
+        assert_identical_grouping_keys(*args)
+        func = op.func
+
+        grouped_df = args[0].obj.groupby(args[0].index)
+        col_names = [col._meta._selected_obj.name for col in args]
+
+        def apply_wrapper(df, apply_func, col_names):
+            cols = (df[col] for col in col_names)
+            return apply_func(*cols)
+
+        # NOTE - We add a detailed meta here so we do not drop the key index
+        # downstream. This seems to be fixed in versions of dask > 2020.12.0
```

Comment:
another reason we should have a higher min :->

Reply:
Yep - I can bump it in the next PR.